File: University/source/Superstition/T3DCreator/Memento.cs
using System;
using System.Drawing;
using System.Collections;

using Microsoft.DirectX;

using FileOperations;

namespace T3DCreator
{
	/// <summary>
	/// Summary description for Memento.
	/// </summary>
	public class Memento
	{
		private ArrayList for_redo, for_undo;
		private FileControl control;

		public Memento()
		{
			this.control = null;
			this.for_redo = new ArrayList();
			this.for_undo = new ArrayList();
		}

		public Memento(FileControl control)
		{
			this.control = control;
			this.for_redo = new ArrayList();
			this.for_undo = new ArrayList();
		}

		public void Push(Operation operation)
		{
			for_undo.Add(operation);
			for_redo.Clear();
			operation.Transform();
			control.Change();
		}

		public void Undo()
		{
			if(for_undo.Count == 0) return;
			int last_index = for_undo.Count - 1;
			Operation operation = for_undo[last_index] as Operation;
			for_undo.RemoveAt(last_index);
			for_redo.Add(operation);
			operation.Invert();
			control.Change();
		}

		public void Redo()
		{
			if(for_redo.Count == 0) return;
			int last_index = for_redo.Count - 1;
			Operation operation = for_redo[last_index] as Operation;
			for_redo.RemoveAt(last_index);
			for_undo.Add(operation);
			operation.Transform();
			control.Change();
		}

		public void Clear()
		{
			for_undo.Clear();
			for_redo.Clear();
		}
	
	} // End of class Memento
}

File: UniversityGameUnported/src/hu/csega/superstition/unported/t3dcreator/Memento.java
package hu.csega.superstition.unported.t3dcreator;

import java.util.ArrayList;
import java.util.List;

import hu.csega.superstition.fileoperations.FileControl;
import hu.csega.superstition.unported.t3dcreator.operations.Operation;

public class Memento {
	private List<Operation> for_redo;
	private List<Operation> for_undo;
	private FileControl control;

	public Memento() {
		this(null);
	}

	public Memento(FileControl control) {
		this.control = control;
		this.for_redo = new ArrayList<>();
		this.for_undo = new ArrayList<>();
	}

	public void Push(Operation operation) {
		for_undo.add(operation);
		for_redo.clear();
		operation.Transform();
		control.change();
	}

	public void Undo() {
		if (for_undo.size() == 0)
			return;

		int last_index = for_undo.size() - 1;
		Operation operation = for_undo.get(last_index);
		for_undo.remove(last_index);
		for_redo.add(operation);
		operation.Invert();
		control.change();
	}

	public void Redo() {
		if (for_redo.size() == 0)
			return;

		int last_index = for_redo.size() - 1;
		Operation operation = for_redo.get(last_index);
		for_redo.remove(last_index);
		for_undo.add(operation);
		operation.Transform();
		control.change();
	}

	public void Clear() {
		for_undo.clear();
		for_redo.clear();
	}

} // End of class Memento